scylla/src/transport/connection.rs

```diff
@@ -1246,31 +1243,48 @@ pub async fn open_named_connection(
-                    let challenge_message = authenticate_challenge.authenticate_message;
-                    unimplemented!(
-                        "Auth Challenge not implemented yet, {:?}",
-                        challenge_message
-                    )
-                }
-                Response::AuthSuccess(_authenticate_success) => {
-                    // OK, continue
-                }
-                Response::Error(err) => {
-                    return Err(err.into());
-                }
-                _ => {
-                    return Err(QueryError::ProtocolError(
-                        "Unexpected response to Authenticate Response message",
-                    ))
-                }
+            match connection.config.authenticator {
+                Some(ref authenticator_provider) => {
+                    let (mut response, mut auth_session) = authenticator_provider
+                        .start_authentication_session(authenticator.clone())
+                        .await
+                        .map_err(QueryError::InvalidMessage)?;
+
+                    loop {
+                        match connection
+                            .authenticate_response(response, authenticator.clone())
+                            .await?.response
+                        {
+                            Response::AuthChallenge(challenge) => {
+                                response = auth_session
+                                    .evaluate_challenge(
+                                        challenge.authenticate_message.as_deref(),
+                                    )
+                                    .await
+                                    .map_err(QueryError::InvalidMessage)?;
+                            }
+                            Response::AuthSuccess(success) => {
+                                auth_session
+                                    .success(success.success_message.as_deref())
+                                    .await
+                                    .map_err(QueryError::InvalidMessage)?;
+                                break;
+                            }
+                            Response::Error(err) => {
+                                return Err(err.into());
+                            }
+                            _ => {
+                                return Err(QueryError::ProtocolError(
+                                    "Unexpected response to Authenticate Response message",
+                                ))
+                            }
+                        }
+                    }
+                },
+                None => return Err(QueryError::InvalidMessage(
+                    "Authentication is required. You can use SessionBuilder::user(\"user\", \"pass\") to provide credentials \
+                    or SessionBuilder::authenticator_provider to provide custom authenticator".to_string(),
+                )),
```

Comment:
The `open_named_connection` function is already very long and is becoming even longer, how about moving the authentication logic to a separate function?